<html><head>
      <meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
   <title>Annotation</title><meta name="generator" content="DocBook XSL Stylesheets V1.70.1"><link rel="start" href="index.html" title="Cytoscape User Manual"><link rel="up" href="index.html" title="Cytoscape User Manual"><link rel="prev" href="ar01s16.html" title="Rendering Engine"><link rel="next" href="ar01s18.html" title="Linkout"></head><body bgcolor="white" text="black" link="#0000FF" vlink="#840084" alink="#0000FF"><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="Annotation"></a>Annotation</h2></div></div></div><p>Annotations in Cytoscape are stored as a set of ontologies (e.g. the Gene Ontology - GO). An ontology consists of a set of controlled vocabulary terms that annotate the objects. For example, using the Gene Ontology, the Saccharomyces Cerevisiae <span class="strong"><strong>CDC55</strong></span> gene&#8217;s biological process is described as &#8220;protein biosynthesis&#8221;, to which GO has assigned the number 6412 (a GO ID).  </p><p></p><pre class="screen">GO 8150 biological_process
 GO 7582 physiological processes
   GO 8152 metabolism
    GO 44238 primary metabolism
      GO 19538 protein metabolism
        GO 6412 protein biosynthesis
</pre><p><span class="strong"><strong>Graphical View of GO Term 6412: protein biosynthesis</strong></span> </p><p><span class="inlinemediaobject"><img src="images/ontology_dag1.png" alt="attachment:ontology_dag1.png"></span> </p><p>Cytoscape can use this ontology DAG (Directed Acyclic Graph) to annotate objects in networks.  The <span class="emphasis"><em><span class="strong"><strong>Ontology Server</strong></span></em></span> (originally called <span class="emphasis"><em><span class="strong"><strong>Bio Data Server</strong></span></em></span>) is a Cytoscape feature which allows you to load, navigate, and assign annotation terms to nodes and edges in a network.  In version 2.4, Cytoscape has enhanced GUI for loading ontology and associated annotation, which enables you to load both local and remote files. </p><p></p><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="Ontology and Annotation File Format"></a>Ontology and Annotation File Format</h3></div></div></div><p>The standard file formats used in Cytoscape Ontology Server are OBO and Gene Association. The GO website details these file formats: </p><div class="itemizedlist"><ul type="disc"><li><p>Ontologies and Definitions:  <a href="http://www.geneontology.org/GO.downloads.shtml#ont" target="_top">http://www.geneontology.org/GO.downloads.shtml#ont</a> </p></li><li><p>Current Annotations: <a href="http://www.geneontology.org/GO.current.annotations.shtml" target="_top">http://www.geneontology.org/GO.current.annotations.shtml</a> </p></li></ul></div><p></p><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a name="OBO File"></a>OBO File</h4></div></div></div><p>OBO file is the ontology DAG itself.  This file defines relationships of ontology terms.  Since version 2.4, Cytoscape can load all ontology files written in OBO format.  The full listing of ontology files are available from Open Biomedical Ontologies (OBO) website:  </p><div class="itemizedlist"><ul type="disc"><li><p>OBO Ontology Browser: <a href="http://obo.sourceforge.net/browse.html" target="_top">http://obo.sourceforge.net/browse.html</a> </p></li></ul></div><p>Sample OBO File - gene_ontology.obo: <a href="http://www.geneontology.org/ontology/gene_ontology_edit.obo" target="_top">http://www.geneontology.org/ontology/gene_ontology_edit.obo</a> </p><pre class="screen">format-version: 1.2
date: 27:11:2006 17:12
saved-by: midori
auto-generated-by: OBO-Edit 1.002
subsetdef: goslim_generic "Generic GO slim"
subsetdef: goslim_goa "GOA and proteome slim"
subsetdef: goslim_plant "Plant GO slim"
subsetdef: goslim_yeast "Yeast GO slim"
subsetdef: gosubset_prok "Prokaryotic GO subset"
default-namespace: gene_ontology
remark: cvs version: $Revision: 5.49 $

[Term]
id: GO:0000001
name: mitochondrion inheritance
namespace: biological_process
def: "The distribution of mitochondria, including the mitochondrial genome, into daughter cells after mitosis or meiosis, mediated by interactions between mitochondria and the cytoskeleton." [GOC:mcc, PMID:10873824, PMID:11389764]
synonym: "mitochondrial inheritance" EXACT []
is_a: GO:0048308 ! organelle inheritance
is_a: GO:0048311 ! mitochondrion distribution

[Term]
id: GO:0000002
name: mitochondrial genome maintenance
namespace: biological_process
def: "The maintenance of the structure and integrity of the mitochondrial genome." [GOC:ai]
is_a: GO:0007005 ! mitochondrion organization and biogenesis
</pre><p><span class="underline"> <span class="strong"><strong>Default List of Ontologies</strong></span> </span> </p><p>Cytoscape provides a list of ontologies available in OBO format.  If Internet connection is available, Cytoscape imports ontology and annotation file directly from the remote source.  The table below is the list of default ontoligies. </p><div class="table"><a name="d0e4488"></a><p class="title"><b>Table&nbsp;21.&nbsp;</b></p><div class="table-contents"><div class="caption"></div><table border="1"><colgroup><col><col></colgroup><tbody><tr><td><p> <span class="strong"><strong>Ontology Name</strong></span> </p></td><td><p> <span class="strong"><strong>Description</strong></span> </p></td></tr><tr><td><p>Gene Ontology Full</p></td><td><p>This data source contains full size GO dag, which contains all GO terms.  This OBO file is written in version 1.2 format.</p></td></tr><tr><td><p>Generic GO slim</p></td><td><p>Subset of general GO Terms.  Includes higer-level terms only.</p></td></tr><tr><td><p>Yeast GO slim</p></td><td><p>Subset of GO Terms for annotating Yeast data sets.  Maintained by SGD.</p></td></tr><tr><td><p>Molecule role (INOH Protein name/family name ontology)</p></td><td><p>A structured controlled vocabulary of concrete protein names and generic (abstract) protein names. This ontology is a INOH pathway annotation ontology, one of a set of ontologies intended to be used in pathway data annotation to ease data integration. This ontology is used to annotate protein names, protein family names, generic/concrete protein names in the INOH pathway data.  INOH is part of the BioPAX working group.</p></td></tr><tr><td><p>Event (INOH pathway ontology)</p></td><td><p>A structured controlled vocabulary of pathway centric biological processes. This ontology is a INOH pathway annotation ontology, one of a set of ontologies intended to be used in pathway data annotation to ease data integration. This ontology is used to annotate biological processes, pathways, sub-pathways in the INOH pathway data. INOH is part of the BioPAX working group.</p></td></tr><tr><td><p>Protein-protein interaction</p></td><td><p>A structured controlled vocabulary for the annotation of experiments concerned with protein-protein interactions.</p></td></tr><tr><td><p>Pathway Ontology</p></td><td><p>The Pathway Ontology is a controlled vocabulary for pathways that provides standard terms for the annotation of geneproducts.</p></td></tr><tr><td><p>PATO</p></td><td><p>PATO is an ontology of phenotypic qualities, intended for use in a number of applications, primarily phenotype annotation.  For more information, please visit PATO wiki (<a href="http://www.bioontology.org/wiki/index.php/PATO:Main_Page" target="_top">http://www.bioontology.org/wiki/index.php/PATO:Main_Page</a>). </p></td></tr><tr><td><p>Mouse pathology</p></td><td><p>Mouse Pathology Ontology (MPATH) is an ontology for mutant mouse pathology.  This is Version 1.</p></td></tr><tr><td><p>Human disease</p></td><td><p>This ontology is a comprehensive hierarchical controlled vocabulary for human disease representation.  For more information, please visit Disease Ontology website (<a href="http://diseaseontology.sourceforge.net/" target="_top">http://diseaseontology.sourceforge.net/</a>).</p></td></tr></tbody></table></div></div><br class="table-break"><p>Although Cytoscape can import all kinds of ontologies in OBO format, <span class="strong"><strong>you need to provide ontology-specific annotation file to annotate node/edge/network in Cytoscape</strong></span>.  For example, while you can annotate human network data with GO Full and Gene Association file for human, you cannot annotate it with the combination of the Human disease ontology file and Gene Association file because the Gene Association file is annotation data only for GO. </p><div class="itemizedlist"><ul type="disc"><li><p>Mapping Ontology - <span class="underline">Annotation files are associated with a specific ontology</span>.  This means, you can map GO terms onto human network using Gene Association file for human, but cannot map Human Disease ontology terms with Gene Association file.  In that case, you need an annotation file for Human disease ontology(see the diagram below): </p></li></ul></div><p><span class="inlinemediaobject"><img src="images/ontology_mapping1.png" alt="attachment:ontology_mapping1.png"></span> </p><p><span class="underline"><span class="strong"><strong>Visualize and Browse Ontology DAG (for Advanced Users)</strong></span></span> </p><p>Relationships between ontology terms are usually represented as Directed Acyclic Graph (DAG).  This is a special case of a network (or graph), where nodes are ontology terms and edges are relationships between terms.  Originally, Cytoscape used special data structure called <a href="http://cytoscape.org/cgi-bin/moin.cgi/BioDataServer" target="_top">BioDataServer</a> to store ontology DAGs.  Since version 2.4, ontology data is stored in the same data structure as normal networks.  This enables users and plugin writers to visualize, browse and manipulate ontology DAGs just like other networks.  The following is an example of visualization of an ontology DAG (Generic GO Slim): </p><p><span class="inlinemediaobject"><img src="images/ontology_dag2.png" alt="attachment:ontology_dag2.png"></span> </p><p>Every ontology term and relationship can have attributes.  In the OBO files, ontology terms have optional fields such as definition, synonyms, comments, or cross-references.  These fields will be imported as node attributes.  To browse those attributes, please use attribute browser (see the example below): </p><p><span class="inlinemediaobject"><img src="images/ontology_attrs.png" alt="attachment:ontology_attrs.png"></span> </p><div class="itemizedlist"><ul type="disc"><li><p>Note 1: Some ontologies have a lot of terms.  For example, the full Gene Ontology contains more than 20,000 terms.  If you need to save memory, you can remove this ontology DAG from Network Panel (Right click on the Ontology Name &#8594; <span class="emphasis"><em><span class="strong"><strong>Destroy Network</strong></span></em></span>). </p></li><li>Note 2: All ontology DAGs will be saved in the session file.  To minimize the session file size, you can delete the Ontology DAG before saving session. </li></ul></div><p></p></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a name="Gene Association File"></a>Gene Association File</h4></div></div></div><p>Gene Association (GA) file is an annotation only for Gene Ontology.  This is a species specific annotation file for GO terms.  <span class="strong"><strong>Gene Assoiation files will only work with Gene Ontologies and NOT others!</strong></span> </p><p>Sample Gene Association File (gene_association.sgd - annotation file for yeast): </p><p></p><pre class="screen">SGD     S000003916      AAD10           GO:0006081      SGD_REF:S000042151|PMID:10572264        ISS             P       aryl-alcohol dehydrogenase (putative)        YJR155W gene    taxon:4932      20020902        SGD
SGD     S000005275      AAD14           GO:0008372      SGD_REF:S000069584      ND              C       aryl-alcohol dehydrogenase (putative)        YNL331C gene    taxon:4932      20010119        SGD
</pre><p></p></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="Import Ontology and Annotation"></a>Import Ontology and Annotation</h3></div></div></div><p><span class="inlinemediaobject"><img src="images/ontology_and_annotation_Import_main.png" alt="attachment:ontology_and_annotation_Import_main.png"></span> Cytoscape 2.4 provides a graphical user interface to import both ontology and annotation files at the same time. </p><p></p><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a name="Import Gene Ontology and Gene Association Files"></a>Import Gene Ontology and Gene Association Files</h4></div></div></div><p>For user's convenience, Cytoscape has a list of URLs for commonly used ontology data and a complete set of Gene Association files.  To import Gene Ontology and Gene Association files for the loaded networks, please follow these steps: </p><p><span class="emphasis"><em><span class="strong"><strong>Important: All data sources in the preset list are remote URLs, meaning a network connection is required!</strong></span></em></span> </p><p><span class="underline">Step 1.</span> Select Annotation File </p><p><span class="inlinemediaobject"><img src="images/ontology_import_annotation.png" alt="attachment:ontology_import_annotation.png"></span> </p><p>Select <span class="underline"><span class="emphasis"><em><span class="strong"><strong>File &#8594; Import &#8594; Ontology and Annotation...</strong></span></em></span></span> to open the <span class="emphasis"><em>Import Ontology and Annotation</em></span> dialog.  From the <span class="emphasis"><em><span class="strong"><strong>Annotation</strong></span></em></span> combo box, select a gene association file for your network.  For example, if you want to annotate yeast network, select <span class="emphasis"><em>Gene Association file for Saccharomyces cerevisiae</em></span>. </p><p><span class="underline">Step 2.</span> Select Ontology File </p><p><span class="inlinemediaobject"><img src="images/ontology_import_obo.png" alt="attachment:ontology_import_obo.png"></span> </p><p>Select an Ontology data (OBO file) from <span class="emphasis"><em><span class="strong"><strong>Ontology</strong></span></em></span> combo box.  If the file is not loaded yet, they are shown in red.  The first three files are Gene Ontology files.  You can load other ontologies, but you need your own annotation file to annotate networks.  </p><p><span class="underline">Step 3.</span> Start Import </p><p>Once you click <span class="emphasis"><em><span class="strong"><strong>Import</strong></span></em></span> button, Cytoscape start loading OBO and Gene Association files from the remote sources.  If you choose <span class="emphasis"><em>GO Full</em></span> it may take a while since it is a large data file. </p><p><span class="underline">Step 4.</span> </p><p>When Cytoscape finishes to import files, the window will be automatically closed.  All attributes mapped by this function have prefix <span class="emphasis"><em>annotation</em></span> and looks like this: <span class="emphasis"><em>annotation</em></span>.[attribute_name].  All ontologies will be added to the end of branch <span class="emphasis"><em><span class="strong"><strong>Ontology DAGs</strong></span></em></span>.  </p><p><span class="inlinemediaobject"><img src="images/ontology_tree.png" alt="attachment:ontology_tree.png"></span> </p><p>Ontology DAGs have some attributes associated with the terms.  All attributes associated with ontology terms have prefix <span class="emphasis"><em><span class="strong"><strong>ontology.</strong></span></em></span>  They have at least one attribute, <span class="emphasis"><em>ontology.name</em></span>.  For more informations about detail of attributes for ontology DAGs, please read <a href="http://www.geneontology.org/GO.format.obo-1_2.shtml" target="_top">OBO specification document</a>. </p><div class="itemizedlist"><ul type="disc"><li>Note: Cytoscape supports both OBO format version 1.0 and 1.2. </li></ul></div><p></p></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="Custom Annotation Files for Ontologies Other than GO (for Advanced Users)"></a>Custom Annotation Files for Ontologies Other than GO (for Advanced Users)</h3></div></div></div><p><span class="emphasis"><em>Ontology and Annotation Import</em></span> function is designed to import general ontology and annotation files.  Internally, mapping ontology terms onto existing networks is same as joining three data tables in Cytoscape.  An Ontology DAG, an annotation file, and networks are used in this process (see the example below). </p><p><span class="underline">Network Data</span> </p><p><span class="inlinemediaobject"><img src="images/ontology_net_table.png" alt="attachment:ontology_net_table.png"></span> </p><p><span class="underline">Ontology Data</span> </p><p><span class="inlinemediaobject"><img src="images/ontology_obo_table.png" alt="attachment:ontology_obo_table.png"></span> </p><p><span class="underline">Annotation Data</span> </p><p><span class="inlinemediaobject"><img src="images/ontology_ga_table.png" alt="attachment:ontology_ga_table.png"></span> </p><p><span class="underline">Mapping Result</span> </p><p><span class="inlinemediaobject"><img src="images/ontology_mapping_result.png" alt="attachment:ontology_mapping_result.png"></span> </p><p>If you want to map ontology terms onto network objects, you need to create a custom annotation file.  The annotation file should contain at least 2 columns: a<span class="emphasis"><em><span class="strong"><strong>primary key</strong></span></em></span> and an <span class="emphasis"><em><span class="strong"><strong>ontology term ID</strong></span></em></span>.  <span class="emphasis"><em>Primary key</em></span> is the value for mapping between annotation file and network.  Usually, node/edge ID is used as primary key, but you may choose any of the available attributes.  <span class="emphasis"><em>Ontology term ID</em></span> is the key for mapping between the ontology DAG and the annotation file.  Using these data sources, you can annotate network objects in Cytoscape. </p><p>Suppose you have a small network: </p><pre class="screen">node_1 pp node_2
node_3 pp node_1
node_2 pp node_3
</pre><p>and you want to annotate this network with <span class="emphasis"><em>Ontology A</em></span>, which is an ontology DAG available in OBO format.  In this case, you need an annotation table file looks like this: </p><pre class="screen">node_1  OA_0000232
node_2  OA_0000441
node_3  OA_0000702
</pre><p>where <span class="emphasis"><em>OA_***</em></span> represents an ontology term ID.  This example is a file with the minimum necessary number of columns, however you can include additional columns that will appear as additional node attributes. </p><p>Some ontologies will be used to annotate edges or networks.  For example, <a href="http://obo.sourceforge.net/cgi-bin/detail.cgi?psi-mi" target="_top">Protein-protein interaction ontology</a> is a controlled set of terms for annotating interactions between proteins, so ontology terms should be mapped onto edges (see example below). </p><pre class="screen">node_1 (pp) node_2  MI:0445
node_3 (pp) node_1  MI:0046
node_2 (pp) node_3  MI:0346
</pre><p><span class="inlinemediaobject"><img src="images/ontology_import_custom1.png" alt="attachment:ontology_import_custom1.png"></span> </p><p>The basic operation of <span class="emphasis"><em>Ontology and Annotation Import</em></span> is the same as <span class="emphasis"><em>Attribute Table Import</em></span>.  The main difference is that you need to specify an additional key for mapping: </p><p><span class="inlinemediaobject"><img src="images/ontology_import_custom2.png" alt="attachment:ontology_import_custom2.png"></span> </p><p>By selecting a column from <span class="emphasis"><em><span class="strong"><strong>Key Column in Annotation File</strong></span></em></span> combo box,  you can specify the key for mapping between ontology terms and annotation file. </p><div class="itemizedlist"><ul type="disc"><li><p>Note: When you load <span class="emphasis"><em>Gene Association</em></span> files, Cytoscape uses a special loader program designed only for Gene Association files.  Because of this program, all attributes will be named automatically.  Also, ontology IDs will be converted into term names and NCBI taxsonomy ID will be converted into actual species name.  However, for custom annotation files, those processes will not be applied.  All ontology terms will be mapped as term IDs. </p></li></ul></div></div></div></body></html>